### ISSUE TITLE
Improve backport friendliness of quotes in Python 3.12 f-string placeholders

### ISSUE THREAD

**Issue by achimnol** [astral-sh/ruff#11056]
I'm maintaining a codebase with multiple release branches targeting different Python versions by release. The latest one uses Python 3.12 while prior versions use Python 3.10 or 3.11.

The recent update of Ruff has introduced Python 3.12 f-string support and the Ruff v0.4 formatter replaces all single-quotes in placeholders to double-quotes.

Example:
`f"a value from dict: {data['key']}"` (the previous way to write string literals in f-string placeholdrs)
→
`f"a value from dict: {data["key"]}"` (allowed in Python 3.12 but a syntax error in Python 3.11 or older)

The problem is that when I **backport** this line to a release branch targeting Python 3.11 or older (but using the same Ruff version), Ruff *silently ignores* the broken syntax. It does not auto-convert the existing codes in Python 3.11, but also does not give errors about the backported codes. Linting will pass, but it will fail when someone executes/imports the code.

In contrast, Black (24.4) loudly fails because the Python parser reports a syntax error.

There could be the following options to resolve the issue:
- Let Ruff 0.4 targeting Python 3.11 or older report the syntax error loudly, just like Black.
- Let Ruff 0.4 targeting Python 3.11 or older auto-convert the double-quotes in f-string placeholders back to single-quotes (or vice-versa depending on the quote style).
- Add an option to keep the single-quotes in f-string placholders intact in Python 3.12 or later.

**Comment by MichaReiser** [comment]
Thanks for the detailed report and we're sorry that you're experiencing this. I think backward compatibility could actually be the reason for changing the default to use single quotes in nested f-strings (@dhruvmanila). 

We have plans to implement lint rules that allow catching unsupported syntax depending on the configured python version. See https://github.com/astral-sh/ruff/issues/6591

**Comment by dhruvmanila** [comment]
Thank you for raising this issue!

> I think backward compatibility could actually be the reason for changing the default to use single quotes in nested f-strings (@dhruvmanila).

Yeah, I think that's a strong reason to not change the quotes and it should be the case until the parser supports target version.

**Comment by dhruvmanila** [comment]
Reading through the suggested solutions, (1) would be done by #6591, I don't think (2) is ideal because that could be complicated.

I think for now it's reasonable to avoid changing the quotes even for 3.12 target version.

Related: https://github.com/astral-sh/ruff/issues/10273, https://github.com/astral-sh/ruff/issues/10184

**Comment by mvaled** [comment]
I consider this a bug if `target-python` is set to py311 or older.

For instance:

```
$ cat src/ruffformatter/__init__.py 
def hello() -> str:
    return f"Hello {"a" + "b"}"


$ cat pyproject.toml 
[project]
name = "ruffformatter"
version = "0.1.0"
description = "Add your description here"
dependencies = []
readme = "README.md"
requires-python = ">= 3.8"
license = { text = "MIT" }

[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[tool.rye]
managed = true
dev-dependencies = [
    "ruff==0.4.2",
]

[tool.hatch.metadata]
allow-direct-references = true

[tool.hatch.build.targets.wheel]
packages = ["src/ruffformatter"]

[tool.ruff]
target-version = "py311"

$ rye run ruff format src/
1 file left unchanged
```

But the file is not syntactically valid in Python 3.11:

```
$ rye run python --version
Python 3.11.7

$ rye run python src/ruffformatter/__init__.py 
  File "/home/manu/tmp/ruffformatter/src/ruffformatter/__init__.py", line 2
    return f"Hello {"a" + "b"}"
                     ^
SyntaxError: f-string: expecting '}'
```

**Comment by MichaReiser** [comment]
@mvaled What I see from the output is that the formatter didn't change the file. The formatter won't change the quote style when targeting python 3.11 or older. That means it's up to you to use the right quotes.

**Comment by charliermarsh** [comment]
Yeah we don't enforce syntax errors right now -- that's known, and isn't a bug but rather a limitation (we just haven't implemented it). What _would_ be a bug is if we _changed_ your code to use syntax that isn't supported by your target version.

**Comment by MichaReiser** [comment]
@dhruvmanila and I talked about this in our 1:1 and I think there are good reasons to changing our current default:

* alternating quotes eases copying code between projects regardless of the target python version
* Alternating quotes is probably what all projects use today and what the community considers to be the standard. Using alternating quotes reduces the diff size
* I do find alternating quote styles easier to parse. It's easier to match the quotes.

**Comment by MichaReiser** [comment]
The exception to the above behavior should be if flipping the quotes reduces the number of necessary escapes:

```python
f"{f"this isn't using single quotes because it contains single quotes that would need escaping"}"}"
```

Keeping the same quotes here should be fine because this syntax is Python 312 or newer

**Comment by MichaReiser** [comment]
Hy @matthewlloyd I'm sorry that you experienced this churn. 

I wont argue about readability. What's more readable is very personal. 

> Seamless copying and pasting: Expressions within f-strings can be cut and pasted freely without needing to adjust quote styles.

I think that should still be given because the formatter will change the quotes for you. 

> Reduced cognitive load. You don't have to remember to switch quotes, it's one less muscle memory to maintain, i

Is this about using nested quotes? Because you can write the f-string with whatever quotes you prefer without worrying if they're the right quotes. The formatter will change the quotes for you.

**Comment by MichaReiser** [comment]
Thanks for the added context.

I'm hesitant about introducing a new setting to configure f-string quotes unless there's a huge demand for it (not saying that your concerns aren't valid).  And I still think that the current behavior is better. For example, changing quotes also breaks syntax highlighting in many tools. But I'm also happy to revisit the default if that's where the community stands. 

I suggest that we create a new issue for this so that we can track this better. @matthewlloyd
